### PR TITLE
persisting execution logs

### PR DESCRIPTION
**Pull request by muhammad-ali-e**
## What

- Implementation of execution log persistence in the database.
- New APIs:
  - List of executions for one workflow
    - Endpoint: GET /workflow/<workflow_id>/execution/
  - List of executions for one workflow
    - Endpoint: GET pipeline/934c29fe-066f-4cad-b78c-97e0b483c9e0/executions/
    - Optional Params:
      - **start_date**=<start_date>
      - **end_date**=<end_date>
      - **page**=<page_number>
      - **page_size**=<page_size>
  - Execution logs
      - Endpoint: GET workflow/execution/<execution_id>/logs/
      - Optional Params:
        - **page**=<page_number>
        - **page_size**=<page_size>      
- Integration with UI added.

## Why

- Essential for tracking and analyzing the history of ETL, TASKs pipelines, and API activities.

## How

- Introduced celery tasks to facilitate logging each activity into the database.

## Possible Improvements

-  Consider inserting logs into the database in batches rather than individually for performance enhancement.

## Database Migrations

- Added a new model for Execution Logs.

## Env Config

- 

## Relevant Docs

-

## Related Issues or PRs

-

## Dependencies Versions

-

## Notes on Testing

- Ensure logs are visible on respective pipeline pages.

## Screenshots

image

image

## Checklist

I have read and understood the [Contribution Guidelines]().

### CODE REVIEW

File: backend/backend/settings/base.py [hunk 1]
```diff
@@ -308,7 +313,7 @@ def get_required_setting(
             "USERNAME": REDIS_USER,
             "PASSWORD": REDIS_PASSWORD,
         },
-        "KEY_FUNCTION": "utils.cache_service.custom_key_function",
+        "KEY_FUNCTION": "utils.redis_cache.custom_key_function",
```
Comment (chandrasekharan-zipstack) [inline backend/backend/settings/base.py:316]: @muhammad-ali-e will this break any existing functionality?

File: backend/utils/cache_service.py [hunk 2]
```diff
@@ -1,8 +1,10 @@
 from typing import Any, Optional
 
-from account.custom_cache import CustomCache
 from django.conf import settings
 from django.core.cache import cache
+from django_redis import get_redis_connection
+
+redis_cache = get_redis_connection("default")
```
Comment (chandrasekharan-zipstack) [inline backend/utils/cache_service.py:7]: NIT: Can this be an instance of the class?

File: backend/utils/constants.py [hunk 3]
```diff
@@ -1,3 +1,14 @@
+import os
+
+from django.conf import settings
+from utils.common_utils import CommonUtils
+
+os.environ.setdefault(
+    "DJANGO_SETTINGS_MODULE",
+    os.environ.get("DJANGO_SETTINGS_MODULE", "backend.settings.dev"),
+)
```
Comment (chandrasekharan-zipstack) [inline backend/utils/constants.py:9]: @muhammad-ali-e why is this required? Ideally django would throw an error if this is not set I think - why do we require to set this here?

File: backend/utils/log_events.py [hunk 4]
```diff
@@ -38,13 +44,98 @@ def disconnect(sid: str) -> None:
...
 
-def _handle_pubsub_messages(message: Any) -> None:
-    channel = message["channel"].decode("utf-8")
-    data = message["data"]
+def _get_validated_log_data(json_data: Any) -> Optional[LogDataDTO]:
+    """Validate log data to persist history
+    Args:
+        json_data (Any): Log data in JSON format
+    """
+    print(f"Received log data: {json_data}")
```
Comment (chandrasekharan-zipstack) [inline backend/utils/log_events.py:52]: @muhammad-ali-e change this to a debug log

File: backend/utils/log_events.py [hunk 5]
```diff
@@ -38,13 +44,98 @@ def disconnect(sid: str) -> None:
...
+    """Store execution log in database
+    Args:
+        data (bytes): Execution log data in bytes format
+    """
+    if not ExecutionLogConstants.IS_ENABLED:
+        return
+    try:
+        log_data = _get_validated_log_data(json_data=data)
+        if log_data:
+            redis_conn.rpush(ExecutionLogConstants.LOG_QUEUE_NAME, log_data.to_json())
```
Comment (chandrasekharan-zipstack) [inline backend/utils/log_events.py:100]: @muhammad-ali-e since we get an `Optional[LogData]` we don't log cases where there is a None returned for some reason - it doesn't get treated as an exception.
It might be better to refactor the function to return `LogData` always instead and throw an exception if its not possible - the exception will be caught in your `catch-all` anyway and logged

File: backend/workflow_manager/workflow/execution.py [hunk 6]
```diff
@@ -305,7 +299,7 @@ def execute_input_file(
         file_name: str,
         single_step: bool,
         file_history: Optional[FileHistory] = None,
-    ) -> None:
+    ) -> bool:
```
Comment (chandrasekharan-zipstack) [inline backend/workflow_manager/workflow/execution.py:302]: NIT: Add a docstring for the return type - I notice that even though a file was executed and returned from cache, we'll return false here. I assume its to indicate whether it was processed or not rather

File: backend/workflow_manager/workflow/execution_log_utils.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            },
+        )
+        if not created:
+            task.enabled = ExecutionLogConstants.IS_ENABLED
+            task.interval = interval
+            task.queue = ExecutionLogConstants.CELERY_QUEUE_NAME
+            task.save()
+            logger.info("Log consumer scheduler updated successfully.")
+        else:
+            print("Log consumer scheduler created successfully.")
```
Comment (chandrasekharan-zipstack) [inline backend/workflow_manager/workflow/execution_log_utils.py:75]: @muhammad-ali-e please change to a log line instead

**Review by chandrasekharan-zipstack — COMMENTED** [review]
Changes seem okay to me for the most part - left some questions / NIT comments

File: backend/utils/dto.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+            json_data = json.loads(json_data)
+            execution_id = json_data.get(LogFieldName.EXECUTION_ID)
+            organization_id = json_data.get(LogFieldName.ORGANIZATION_ID)
+            timestamp = json_data.get(LogFieldName.TIMESTAMP)
+            log_type = json_data.get(LogFieldName.TYPE)
+            data = json_data.get(LogFieldName.DATA)
+            if all((execution_id, organization_id, timestamp, log_type, data)):
+                return cls(execution_id, organization_id, timestamp, log_type, data)
+        except (json.JSONDecodeError, AttributeError):
+            pass
```
Comment (vishnuszipstack) [inline backend/utils/dto.py:48]: @muhammad-ali-e @athul-rs can we add debug log here, will be helpful to debug something happens here.

File: backend/backend/settings/base.py (hunk 1 above)
Comment (muhammad-ali-e) [reply backend/backend/settings/base.py:316]: I don't think it breaks, It just move the function from one class/file  to another

File: backend/utils/cache_service.py (hunk 2 above)
Comment (muhammad-ali-e) [reply backend/utils/cache_service.py:7]: I think it's better to put in outside of class , There's no need to create a new instance of the connection for each instance of CacheService.

File: backend/utils/constants.py (hunk 3 above)
Comment (muhammad-ali-e) [reply backend/utils/constants.py:9]: The log_events function relies on constants from constants.py. Since this script runs outside the Django context, we need to manually initialize the Django settings module to ensure the necessary configuration is available like we had in celery.py, wsgi.py etc..

File: backend/utils/dto.py (hunk 8 above)
Comment (muhammad-ali-e) [reply backend/utils/dto.py]: addressed

File: backend/workflow_manager/workflow/execution.py (hunk 6 above)
Comment (muhammad-ali-e) [reply backend/workflow_manager/workflow/execution.py:302]: addressed

File: backend/utils/log_events.py (hunk 5 above)
Comment (muhammad-ali-e) [reply backend/utils/log_events.py:107]: addressed by adding more loggers in that function

**Review by vishnuszipstack — APPROVED** [review]

File: backend/workflow_manager/workflow/execution_log_utils.py (hunk 7 above)
Comment (muhammad-ali-e) [reply backend/workflow_manager/workflow/execution_log_utils.py]: UPDATED

File: backend/utils/log_events.py (hunk 4 above)
Comment (muhammad-ali-e) [reply backend/utils/log_events.py]: UPDATED

**Review by chandrasekharan-zipstack — APPROVED** [review]
LGTM - minor concern around setting an env in `constants.py` for DJANGO_SETTINGS_MODULE alone

File: backend/utils/cache_service.py (hunk 2 above)
Comment (chandrasekharan-zipstack) [reply backend/utils/cache_service.py:7]: I think it can be added as a class attribute to ensure there's only one connection for all objects of the class - but yeah we can keep it in global scope until we face an issue

**Review by hari-kuriakose — APPROVED** [review]
@muhammad-ali-e Migration looks ok.
